internal/kms/keyprotect.go

```diff
@@ -0,0 +1,234 @@
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = KeyProtectMetadataDefaultSecretsName
+	}
+
+	err = setConfigString(&kms.serviceInstanceID, args.Config, keyProtectServiceInstanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// read the Kubernetes Secret with credentials
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets for %T: %w", kms,
+			err)
+	}
+
+	err = setConfigString(&kms.serviceAPIKey, secrets, KeyProtectServiceAPIKey)
+	if err != nil {
+		return nil, err
+	}
+	err = setConfigString(&kms.customerRootKey, secrets, KeyProtectCustomerRootKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// KeyProtectSessionToken is optional
+	err = setConfigString(&kms.sessionToken, secrets, KeyProtectSessionToken)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+
+	// KeyProtect Region is optional
+	err = setConfigString(&kms.region, args.Config, KeyProtectRegionKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+
+	// crk arn is optional
+	err = setConfigString(&kms.crk, secrets, KeyProtectCRK)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+
+	return kms, nil
+}
+
+func (kms *KeyProtectKMS) getSecrets() (map[string]interface{}, error) {
+	c, err := k8s.NewK8sClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to Kubernetes to "+
+			"get Secret %s/%s: %w", kms.namespace, kms.secretName, err)
+	}
+
+	secret, err := c.CoreV1().Secrets(kms.namespace).Get(context.TODO(),
+		kms.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Secret %s/%s: %w",
+			kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]interface{})
+
+	for k, v := range secret.Data {
+		switch k {
+		case KeyProtectServiceAPIKey, KeyProtectCustomerRootKey, KeyProtectSessionToken, KeyProtectCRK:
+			config[k] = string(v)
+		default:
+			return nil, fmt.Errorf("unsupported option for KMS "+
+				"provider %q: %s", kmsTypeKeyProtectMetadata, k)
+		}
+	}
+
+	return config, nil
+}
+
+func (kms *KeyProtectKMS) Destroy() {
+	// Nothing to do.
+}
+
+func (kms *KeyProtectKMS) RequiresDEKStore() DEKStoreType {
+	return DEKStoreMetadata
+}
+
+func (kms *KeyProtectKMS) getService() error {
+	// Use your Service API Key and your KeyProtect Service Instance ID to create a ClientConfig
+	cc := kp.ClientConfig{
+		BaseURL:    kp.DefaultBaseURL,
+		APIKey:     kms.serviceAPIKey,
+		InstanceID: kms.serviceInstanceID,
+	}
+
+	// Build a new client from the config
+	client, err := kp.New(cc, kp.DefaultTransport())
+	if err != nil {
+		return fmt.Errorf("failed to create keyprotect client: %w", err)
+	}
+
+	// List keys in the KeyProtect instance
+	_, err = client.GetKeys(context.Background(), 0, 0)
+	if err != nil {
+		return fmt.Errorf("unable to fetch keys from key protect")
+	}
```

Comment:
why this is needed? if needed why err is not returned back?

Reply:
it make sure we have the rbac..etc required for operating with the KMS server. I have returned that error too already?

Reply:
I don't think it's a valid reason. the RBAC errors will be returned during create operation. please remove this one and you are not returning/wrapping the original error message